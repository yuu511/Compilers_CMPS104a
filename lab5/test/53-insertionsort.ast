ROOT "<<ROOT>>" 0.0.0 {0} 
FUNCTION "" 6.22.1 {0} void function 
TYPE_ID "" 6.22.1 {0} 
VOID "void" 6.22.1 {0} 
IDENT "fail" 6.22.6 {0} 
PARAM "(" 6.22.11 {0} 
TYPE_ID "" 6.22.12 {1} string variable param lval 
STRING "string" 6.22.12 {0} 
IDENT "expr" 6.22.19 {0} 
TYPE_ID "" 6.22.25 {1} string variable param lval 
STRING "string" 6.22.25 {0} 
IDENT "file" 6.22.32 {0} 
TYPE_ID "" 6.22.38 {1} int variable param lval 
INT "int" 6.22.38 {0} 
IDENT "line" 6.22.42 {0} 
FUNCTION "" 6.24.1 {1} void function 
TYPE_ID "" 6.24.1 {0} 
VOID "void" 6.24.1 {0} 
IDENT "putchr" 6.24.6 {0} 
PARAM "(" 6.24.13 {0} 
TYPE_ID "" 6.24.14 {2} int variable param lval 
INT "int" 6.24.14 {0} 
IDENT "chr" 6.24.18 {0} 
FUNCTION "" 6.25.1 {2} void function 
TYPE_ID "" 6.25.1 {0} 
VOID "void" 6.25.1 {0} 
IDENT "putint" 6.25.6 {0} 
PARAM "(" 6.25.13 {0} 
TYPE_ID "" 6.25.14 {3} int variable param lval 
INT "int" 6.25.14 {0} 
IDENT "num" 6.25.18 {0} 
FUNCTION "" 6.26.1 {3} void function 
TYPE_ID "" 6.26.1 {0} 
VOID "void" 6.26.1 {0} 
IDENT "putstr" 6.26.6 {0} 
PARAM "(" 6.26.13 {0} 
TYPE_ID "" 6.26.14 {4} string variable param lval 
STRING "string" 6.26.14 {0} 
IDENT "str" 6.26.21 {0} 
FUNCTION "" 6.28.1 {4} int function 
TYPE_ID "" 6.28.1 {0} 
INT "int" 6.28.1 {0} 
IDENT "getchr" 6.28.5 {0} 
PARAM "(" 6.28.11 {0} 
FUNCTION "" 6.29.1 {5} string function 
TYPE_ID "" 6.29.1 {0} 
STRING "string" 6.29.1 {0} 
IDENT "getstr" 6.29.8 {0} 
PARAM "(" 6.29.14 {0} 
FUNCTION "" 6.30.1 {6} string function 
TYPE_ID "" 6.30.1 {0} 
STRING "string" 6.30.1 {0} 
IDENT "getln" 6.30.8 {0} 
PARAM "(" 6.30.13 {0} 
FUNCTION "" 7.8.1 {7} int function 
TYPE_ID "" 7.8.1 {0} 
INT "int" 7.8.1 {0} 
IDENT "strcmp" 7.8.5 {0} 
PARAM "(" 7.8.12 {0} 
TYPE_ID "" 7.8.13 {8} string variable param lval 
STRING "string" 7.8.13 {0} 
IDENT "s1" 7.8.20 {0} 
TYPE_ID "" 7.8.24 {8} string variable param lval 
STRING "string" 7.8.24 {0} 
IDENT "s2" 7.8.31 {0} 
BLOCK "{" 7.8.35 {8} 
TYPE_ID "" 7.9.4 {8} int variable local lval vreg (7.9.4)
INT "int" 7.9.4 {0} 
IDENT "index" 7.9.8 {0} 
INTCON "0" 7.9.16 {8} int variable local lval vreg 
TYPE_ID "" 7.10.4 {8} int variable local lval vreg (7.10.4)
INT "int" 7.10.4 {0} 
IDENT "contin" 7.10.8 {0} 
INTCON "1" 7.10.17 {8} int variable local lval vreg 
TYPE_ID "" 7.11.4 {8} int variable local lval vreg (7.11.4)
INT "int" 7.11.4 {0} 
IDENT "s1c" 7.11.8 {0} 
INTCON "0" 7.11.14 {8} int variable local lval vreg 
TYPE_ID "" 7.12.4 {8} int variable local lval vreg (7.12.4)
INT "int" 7.12.4 {0} 
IDENT "s2c" 7.12.8 {0} 
INTCON "0" 7.12.14 {8} int variable local lval vreg 
TYPE_ID "" 7.13.4 {8} int variable local lval vreg (7.13.4)
INT "int" 7.13.4 {0} 
IDENT "cmp" 7.13.8 {0} 
INTCON "0" 7.13.14 {8} int variable local lval vreg 
WHILE "while" 7.14.4 {8} 
IDENT "contin" 7.14.11 {8} int variable local lval (7.10.4)
BLOCK "{" 7.14.19 {8} 
'=' "=" 7.15.11 {8} int variable local lval vaddr 
IDENT "s1c" 7.15.7 {8} int variable local lval (7.11.4)
INDEX "[" 7.15.15 {8} int lval vaddr 
IDENT "s1" 7.15.13 {8} string variable param lval (7.8.13)
IDENT "index" 7.15.16 {8} int variable local lval (7.9.4)
'=' "=" 7.16.11 {8} int variable local lval vaddr 
IDENT "s2c" 7.16.7 {8} int variable local lval (7.12.4)
INDEX "[" 7.16.15 {8} int lval vaddr 
IDENT "s2" 7.16.13 {8} string variable param lval (7.8.24)
IDENT "index" 7.16.16 {8} int variable local lval (7.9.4)
'=' "=" 7.17.11 {8} int variable local lval vreg 
IDENT "cmp" 7.17.7 {8} int variable local lval (7.13.4)
'-' "-" 7.17.17 {8} int vreg 
IDENT "s1c" 7.17.13 {8} int variable local lval (7.11.4)
IDENT "s2c" 7.17.19 {8} int variable local lval (7.12.4)
IF "if" 7.18.7 {8} 
NE "!=" 7.18.15 {8} int vreg 
IDENT "cmp" 7.18.11 {8} int variable local lval (7.13.4)
INTCON "0" 7.18.18 {8} int const 
RETURN "return" 7.18.21 {8} 
IDENT "cmp" 7.18.28 {8} int variable local lval (7.13.4)
IF "if" 7.19.7 {8} 
EQ "==" 7.19.15 {8} int vreg 
IDENT "s1c" 7.19.11 {8} int variable local lval (7.11.4)
CHARCON "'\0'" 7.19.18 {8} int const 
'=' "=" 7.19.31 {8} int variable local lval vreg 
IDENT "contin" 7.19.24 {8} int variable local lval (7.10.4)
INTCON "0" 7.19.33 {8} int const 
IF "if" 7.20.7 {8} 
EQ "==" 7.20.15 {8} int vreg 
IDENT "s2c" 7.20.11 {8} int variable local lval (7.12.4)
CHARCON "'\0'" 7.20.18 {8} int const 
'=' "=" 7.20.31 {8} int variable local lval vreg 
IDENT "contin" 7.20.24 {8} int variable local lval (7.10.4)
INTCON "0" 7.20.33 {8} int const 
'=' "=" 7.21.13 {8} int variable local lval vreg 
IDENT "index" 7.21.7 {8} int variable local lval (7.9.4)
'+' "+" 7.21.21 {8} int vreg 
IDENT "index" 7.21.15 {8} int variable local lval (7.9.4)
INTCON "1" 7.21.23 {8} int const 
RETURN "return" 7.23.4 {8} 
INTCON "0" 7.23.11 {8} int const 
FUNCTION "" 7.26.1 {8} void function 
TYPE_ID "" 7.26.1 {0} 
VOID "void" 7.26.1 {0} 
IDENT "insertion_sort" 7.26.6 {0} 
PARAM "(" 7.26.21 {0} 
TYPE_ID "" 7.26.22 {9} int variable param lval 
INT "int" 7.26.22 {0} 
IDENT "size" 7.26.26 {0} 
TYPE_ID "" 7.26.32 {9} string array variable param lval 
ARRAY "array" 7.26.32 {0} 
STRING "string" 7.26.38 {0} 
IDENT "words" 7.26.46 {0} 
BLOCK "{" 7.26.53 {9} 
TYPE_ID "" 7.27.4 {9} int variable local lval vreg (7.27.4)
INT "int" 7.27.4 {0} 
IDENT "sorted" 7.27.8 {0} 
INTCON "1" 7.27.17 {9} int variable local lval vreg 
TYPE_ID "" 7.28.4 {9} int variable local lval vreg (7.28.4)
INT "int" 7.28.4 {0} 
IDENT "slot" 7.28.8 {0} 
INTCON "0" 7.28.15 {9} int variable local lval vreg 
TYPE_ID "" 7.29.4 {9} string variable local lval vreg (7.29.4)
STRING "string" 7.29.4 {0} 
IDENT "element" 7.29.11 {0} 
NULLPTR "nullptr" 7.29.21 {9} string variable local lval vreg 
TYPE_ID "" 7.30.4 {9} int variable local lval vreg (7.30.4)
INT "int" 7.30.4 {0} 
IDENT "contin" 7.30.8 {0} 
INTCON "0" 7.30.17 {9} int variable local lval vreg 
WHILE "while" 7.31.4 {9} 
LT "<" 7.31.18 {9} int vreg 
IDENT "sorted" 7.31.11 {9} int variable local lval (7.27.4)
IDENT "size" 7.31.20 {9} int variable param lval (7.26.22)
BLOCK "{" 7.31.26 {9} 
'=' "=" 7.32.12 {9} int variable local lval vreg 
IDENT "slot" 7.32.7 {9} int variable local lval (7.28.4)
IDENT "sorted" 7.32.14 {9} int variable local lval (7.27.4)
'=' "=" 7.33.15 {9} string variable local lval vaddr 
IDENT "element" 7.33.7 {9} string variable local lval (7.29.4)
INDEX "[" 7.33.22 {9} string lval vaddr 
IDENT "words" 7.33.17 {9} string array variable param lval (7.26.32)
IDENT "slot" 7.33.23 {9} int variable local lval (7.28.4)
'=' "=" 7.34.14 {9} int variable local lval vreg 
IDENT "contin" 7.34.7 {9} int variable local lval (7.30.4)
INTCON "1" 7.34.16 {9} int const 
WHILE "while" 7.35.7 {9} 
IDENT "contin" 7.35.14 {9} int variable local lval (7.30.4)
BLOCK "{" 7.35.22 {9} 
IF "if" 7.36.10 {9} 
EQ "==" 7.36.19 {9} int vreg 
IDENT "slot" 7.36.14 {9} int variable local lval (7.28.4)
INTCON "0" 7.36.22 {9} int const 
BLOCK "{" 7.36.25 {9} 
'=' "=" 7.37.20 {9} int variable local lval vreg 
IDENT "contin" 7.37.13 {9} int variable local lval (7.30.4)
INTCON "0" 7.37.22 {9} int const 
IF "if" 7.38.16 {9} 
LE "<=" 7.38.54 {9} int vreg 
CALL "(" 7.38.27 {7} int function (7.8.1)
IDENT "strcmp" 7.38.20 {0} 
INDEX "[" 7.38.33 {9} string lval vaddr 
IDENT "words" 7.38.28 {9} string array variable param lval (7.26.32)
'-' "-" 7.38.39 {9} int vreg 
IDENT "slot" 7.38.34 {9} int variable local lval (7.28.4)
INTCON "1" 7.38.41 {9} int const 
IDENT "element" 7.38.45 {9} string variable local lval (7.29.4)
INTCON "0" 7.38.57 {9} int const 
BLOCK "{" 7.38.60 {9} 
'=' "=" 7.39.20 {9} int variable local lval vreg 
IDENT "contin" 7.39.13 {9} int variable local lval (7.30.4)
INTCON "0" 7.39.22 {9} int const 
BLOCK "{" 7.40.16 {9} 
'=' "=" 7.41.25 {9} string lval vaddr 
INDEX "[" 7.41.18 {9} string lval vaddr 
IDENT "words" 7.41.13 {9} string array variable param lval (7.26.32)
IDENT "slot" 7.41.19 {9} int variable local lval (7.28.4)
INDEX "[" 7.41.32 {9} string lval vaddr 
IDENT "words" 7.41.27 {9} string array variable param lval (7.26.32)
'-' "-" 7.41.38 {9} int vreg 
IDENT "slot" 7.41.33 {9} int variable local lval (7.28.4)
INTCON "1" 7.41.40 {9} int const 
'=' "=" 7.42.18 {9} int variable local lval vreg 
IDENT "slot" 7.42.13 {9} int variable local lval (7.28.4)
'-' "-" 7.42.25 {9} int vreg 
IDENT "slot" 7.42.20 {9} int variable local lval (7.28.4)
INTCON "1" 7.42.27 {9} int const 
'=' "=" 7.45.19 {9} string lval vreg vaddr 
INDEX "[" 7.45.12 {9} string lval vaddr 
IDENT "words" 7.45.7 {9} string array variable param lval (7.26.32)
IDENT "slot" 7.45.13 {9} int variable local lval (7.28.4)
IDENT "element" 7.45.21 {9} string variable local lval (7.29.4)
'=' "=" 7.46.14 {9} int variable local lval vreg 
IDENT "sorted" 7.46.7 {9} int variable local lval (7.27.4)
'+' "+" 7.46.23 {9} int vreg 
IDENT "sorted" 7.46.16 {9} int variable local lval (7.27.4)
INTCON "1" 7.46.25 {9} int const 
FUNCTION "" 7.51.1 {9} void function 
TYPE_ID "" 7.51.1 {0} 
VOID "void" 7.51.1 {0} 
IDENT "print_array" 7.51.6 {0} 
PARAM "(" 7.51.18 {0} 
TYPE_ID "" 7.51.19 {10} int variable param lval 
INT "int" 7.51.19 {0} 
IDENT "size" 7.51.23 {0} 
TYPE_ID "" 7.51.29 {10} string array variable param lval 
ARRAY "array" 7.51.29 {0} 
STRING "string" 7.51.35 {0} 
IDENT "words" 7.51.43 {0} 
BLOCK "{" 7.51.50 {10} 
TYPE_ID "" 7.52.4 {10} int variable local lval vreg (7.52.4)
INT "int" 7.52.4 {0} 
IDENT "index" 7.52.8 {0} 
INTCON "0" 7.52.16 {10} int variable local lval vreg 
WHILE "while" 7.53.4 {10} 
LT "<" 7.53.17 {10} int vreg 
IDENT "index" 7.53.11 {10} int variable local lval (7.52.4)
IDENT "size" 7.53.19 {10} int variable param lval (7.51.19)
BLOCK "{" 7.53.25 {10} 
CALL "(" 7.54.14 {3} void function (6.26.1)
IDENT "putstr" 7.54.7 {0} 
INDEX "[" 7.54.20 {10} string lval vaddr 
IDENT "words" 7.54.15 {10} string array variable param lval (7.51.29)
IDENT "index" 7.54.21 {10} int variable local lval (7.52.4)
CALL "(" 7.55.14 {1} void function (6.24.1)
IDENT "putchr" 7.55.7 {0} 
CHARCON "'\n'" 7.55.15 {10} int const 
'=' "=" 7.56.13 {10} int variable local lval vreg 
IDENT "index" 7.56.7 {10} int variable local lval (7.52.4)
'+' "+" 7.56.21 {10} int vreg 
IDENT "index" 7.56.15 {10} int variable local lval (7.52.4)
INTCON "1" 7.56.23 {10} int const 
FUNCTION "" 7.60.1 {10} int function 
TYPE_ID "" 7.60.1 {0} 
INT "int" 7.60.1 {0} 
IDENT "read_words" 7.60.5 {0} 
PARAM "(" 7.60.16 {0} 
TYPE_ID "" 7.60.17 {11} int variable param lval 
INT "int" 7.60.17 {0} 
IDENT "size" 7.60.21 {0} 
TYPE_ID "" 7.60.27 {11} string array variable param lval 
ARRAY "array" 7.60.27 {0} 
STRING "string" 7.60.33 {0} 
IDENT "words" 7.60.41 {0} 
BLOCK "{" 7.60.48 {11} 
TYPE_ID "" 7.61.4 {11} int variable local lval vreg (7.61.4)
INT "int" 7.61.4 {0} 
IDENT "count" 7.61.8 {0} 
INTCON "0" 7.61.16 {11} int variable local lval vreg 
TYPE_ID "" 7.62.4 {11} string variable local lval vreg (7.62.4)
STRING "string" 7.62.4 {0} 
IDENT "word" 7.62.11 {0} 
NULLPTR "nullptr" 7.62.18 {11} string variable local lval vreg 
WHILE "while" 7.63.4 {11} 
INTCON "1" 7.63.11 {11} int const 
BLOCK "{" 7.63.14 {11} 
IF "if" 7.64.7 {11} 
EQ "==" 7.64.17 {11} int vreg 
IDENT "count" 7.64.11 {11} int variable local lval (7.61.4)
IDENT "size" 7.64.20 {11} int variable param lval (7.60.17)
RETURN "return" 7.64.26 {11} 
IDENT "count" 7.64.33 {11} int variable local lval (7.61.4)
'=' "=" 7.65.12 {11} string variable local lval vreg 
IDENT "word" 7.65.7 {11} string variable local lval (7.62.4)
CALL "(" 7.65.20 {5} string function (6.29.1)
IDENT "getstr" 7.65.14 {0} 
IF "if" 7.66.7 {11} 
EQ "==" 7.66.16 {11} int vreg 
IDENT "word" 7.66.11 {11} string variable local lval (7.62.4)
NULLPTR "nullptr" 7.66.19 {11} nullptr_t const 
RETURN "return" 7.66.28 {11} 
IDENT "count" 7.66.35 {11} int variable local lval (7.61.4)
'=' "=" 7.67.20 {11} string lval vreg vaddr 
INDEX "[" 7.67.12 {11} string lval vaddr 
IDENT "words" 7.67.7 {11} string array variable param lval (7.60.27)
IDENT "count" 7.67.13 {11} int variable local lval (7.61.4)
IDENT "word" 7.67.22 {11} string variable local lval (7.62.4)
'=' "=" 7.68.13 {11} int variable local lval vreg 
IDENT "count" 7.68.7 {11} int variable local lval (7.61.4)
'+' "+" 7.68.21 {11} int vreg 
IDENT "count" 7.68.15 {11} int variable local lval (7.61.4)
INTCON "1" 7.68.23 {11} int const 
FUNCTION "" 7.72.1 {11} int function 
TYPE_ID "" 7.72.1 {0} 
INT "int" 7.72.1 {0} 
IDENT "main" 7.72.5 {0} 
PARAM "(" 7.72.9 {0} 
BLOCK "{" 7.72.12 {12} 
TYPE_ID "" 7.73.4 {12} int variable local lval vreg (7.73.4)
INT "int" 7.73.4 {0} 
IDENT "count" 7.73.8 {0} 
INTCON "100" 7.73.16 {12} int variable local lval vreg 
TYPE_ID "" 7.74.4 {12} string array variable local lval vreg (7.74.4)
ARRAY "array" 7.74.4 {0} 
STRING "string" 7.74.10 {0} 
IDENT "words" 7.74.18 {0} 
ALLOC "alloc" 7.74.26 {12} string array variable local lval vreg 
ARRAY "array" 7.74.32 {0} 
STRING "string" 7.74.38 {0} 
IDENT "count" 7.74.47 {12} int variable local lval (7.73.4)
'=' "=" 7.75.10 {12} int variable local lval vreg 
IDENT "count" 7.75.4 {12} int variable local lval (7.73.4)
CALL "(" 7.75.23 {10} int function (7.60.1)
IDENT "read_words" 7.75.12 {0} 
IDENT "count" 7.75.24 {12} int variable local lval (7.73.4)
IDENT "words" 7.75.31 {12} string array variable local lval (7.74.4)
CALL "(" 7.76.19 {8} void function (7.26.1)
IDENT "insertion_sort" 7.76.4 {0} 
IDENT "count" 7.76.20 {12} int variable local lval (7.73.4)
IDENT "words" 7.76.27 {12} string array variable local lval (7.74.4)
CALL "(" 7.77.16 {9} void function (7.51.1)
IDENT "print_array" 7.77.4 {0} 
IDENT "count" 7.77.17 {12} int variable local lval (7.73.4)
IDENT "words" 7.77.24 {12} string array variable local lval (7.74.4)
RETURN "return" 7.78.4 {12} 
INTCON "0" 7.78.11 {12} int const 
