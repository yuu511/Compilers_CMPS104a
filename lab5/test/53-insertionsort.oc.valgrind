==24298== Memcheck, a memory error detector
==24298== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==24298== Using Valgrind-3.14.0 and LibVEX; rerun with -h for copyright info
==24298== Command: ../oc 53-insertionsort.oc
==24298== 
==24298== 
==24298== HEAP SUMMARY:
==24298==     in use at exit: 0 bytes in 0 blocks
==24298==   total heap usage: 1,809 allocs, 1,809 frees, 230,831 bytes allocated
==24298== 
==24298== All heap blocks were freed -- no leaks are possible
==24298== 
==24298== For counts of detected and suppressed errors, rerun with: -v
==24298== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
