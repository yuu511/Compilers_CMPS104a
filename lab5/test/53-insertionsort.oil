strcmp:    .function int 
           .param ptr s1
           .param ptr s2
           .local int index
           .local int contin
           .local int s1c
           .local int s2c
           .local int cmp
           index = 0  
           contin = 1  
           s1c = 0  
           s2c = 0  
           cmp = 0  
.wh0:      $t0:i = contin  
           goto .od0 if not $t0:i
.do0:      $t1:i = s1[index * :i]  
           s1c = $t1:i  
           $t2:i = s2[index * :i]  
           s2c = $t2:i  
           $t3:i = s1c - s2c
           cmp = $t3:i  
.if0:      $t4:i = cmp != 0
           goto .fi0 if not $t4:i
.th0:      return cmp
.fi0:
.if1:      $t5:i = s1c == '\0'
           goto .fi1 if not $t5:i
.th1:      contin = 0  
.fi1:
.if2:      $t6:i = s2c == '\0'
           goto .fi2 if not $t6:i
.th2:      contin = 0  
.fi2:
           $t7:i = index + 1
           index = $t7:i  
           goto .wh0
.od0:
           return 0
           return
           .end
insertion_sort: .function 
           .param int size
           .param ptr words
           .local int sorted
           .local int slot
           .local ptr element
           .local int contin
           sorted = 1  
           slot = 0  
           element = nullptr  
           contin = 0  
.wh1:      $t0:i = sorted < size
           goto .od1 if not $t0:i
.do1:      slot = sorted  
           $t1:c = words[slot * :c]  
           element = $t1:c  
           contin = 1  
.wh2:      $t2:i = contin  
           goto .od2 if not $t2:i
.do2:
.if3:      $t3:i = slot == 0
           goto .el3 if not $t3:i
.th3:      contin = 0  
           goto .fi3
.el3:
.if4:      $t7:i = slot - 1
           $t6:c = words[$t7:i * :c]  
           $t5:i = call strcmp($t6:c,element)  
           $t4:i = $t5:i <= 0
           goto .el4 if not $t4:i
.th4:      contin = 0  
           goto .fi4
.el4:      $t9:i = slot - 1
           $t8:c = words[$t9:i * :c]  
           words[slot * :c] = $t8:c  
           $t10:i = slot - 1
           slot = $t10:i  
.fi4:
.fi3:
           goto .wh2
.od2:
           words[slot * :c] = element  
           $t11:i = sorted + 1
           sorted = $t11:i  
           goto .wh1
.od1:
           return
           .end
print_array: .function 
           .param int size
           .param ptr words
           .local int index
           index = 0  
.wh3:      $t0:i = index < size
           goto .od3 if not $t0:i
.do3:      $t1:c = words[index * :c]  
           call putstr($t1:c)  
           call putchr('\n')  
           $t2:i = index + 1
           index = $t2:i  
           goto .wh3
.od3:
           return
           .end
read_words: .function int 
           .param int size
           .param ptr words
           .local int count
           .local ptr word
           count = 0  
           word = nullptr  
.wh4:      $t0:i = 1  
           goto .od4 if not $t0:i
.do4:
.if5:      $t1:i = count == size
           goto .fi5 if not $t1:i
.th5:      return count
.fi5:
           $t2:p = call getstr()  
           word = $t2:p  
.if6:      $t3:i = word == nullptr
           goto .fi6 if not $t3:i
.th6:      return count
.fi6:
           words[count * :c] = word  
           $t4:i = count + 1
           count = $t4:i  
           goto .wh4
.od4:
           return
           .end
main:      .function int 
           .local int count
           .local ptr words
           count = 100  
           $t0:p = count * sizeof ptr
           words = call malloc($t0:p)  
           $t1:i = call read_words(count,words)  
           count = $t1:i  
           call insertion_sort(count,words)  
           call print_array(count,words)  
           return 0
           return
           .end
