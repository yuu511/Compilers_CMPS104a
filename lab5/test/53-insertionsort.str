string_set[   0]:    7296690267450279874 0x55b6f79b6d58->"chr"
                   13348923631108115385 0x55b6f79b7718->"putstr"
string_set[   4]:    7835397169562667056 0x55b6f79bbe48->"words"
string_set[   5]:   12679825168387489523 0x55b6f79b7278->"num"
string_set[   6]:    1782657156151550167 0x55b6f79c0c58->"count"
                    4851023869765522817 0x55b6f79bc048->"sorted"
                    9317282556992397945 0x55b6f79ba718->"if"
string_set[   7]:    6814028903552918164 0x55b6f79bbbe8->"array"
                    2297668033614959926 0x55b6f79b8d78->"0"
string_set[  10]:    1583576014053361350 0x55b6f79b78d8->"str"
                   13497578857207247412 0x55b6f79b6318->"string"
                   11597697714117577063 0x55b6f79b8cb8->"="
string_set[  11]:    1396094175042499165 0x55b6f79bbca8->"<"
string_set[  13]:   12868445110721718657 0x55b6f79b6978->")"
string_set[  16]:   11911796782241914524 0x55b6f79b5088->"void"
                    4698864730501174006 0x55b6f79b8bf8->"index"
                   10159970873491820195 0x55b6f79b9018->"1"
string_set[  17]:    3950881983657090583 0x55b6f79bf1b8->"print_array"
string_set[  18]:   13581035385482130682 0x55b6f79b9d38->"["
string_set[  19]:    3730725358860599826 0x55b6f79bb868->"insertion_sort"
                    5344662657899890615 0x55b6f79b6a38->";"
string_set[  20]:   11868120863286193964 0x55b6f79c2278->"main"
                   15816506932648033945 0x55b6f79b6258->"("
string_set[  21]:    2617056156435215158 0x55b6f79ba8d8->"!="
string_set[  23]:    6340238197210063670 0x55b6f79c2638->"100"
                    7629915659344845568 0x55b6f79b8938->"s2"
string_set[  24]:    3680345365606771475 0x55b6f79bc568->"element"
                   13934964484049229816 0x55b6f79bba28->"size"
string_set[  25]:   11166701219533395393 0x55b6f79b7ef8->"getstr"
string_set[  27]:   12808232827189627082 0x55b6f79bc6a8->"nullptr"
string_set[  28]:   14601455746388216058 0x55b6f79b86d8->"s1"
string_set[  29]:    1370730858159036685 0x55b6f79bb528->"+"
                   14635384904416252381 0x55b6f79baa98->"return"
                   12677793202873405033 0x55b6f79b91a8->"s1c"
string_set[  34]:    8290819294854170237 0x55b6f79c2928->"alloc"
                    7211130878751977441 0x55b6f79b6b98->"putchr"
string_set[  35]:   14736352883882276958 0x55b6f79bac98->"=="
string_set[  36]:   11752130776608228127 0x55b6f79bad58->"'\0'"
string_set[  37]:    6192799629375248351 0x55b6f79c0158->"'\n'"
                   12291392672316406109 0x55b6f79b6898->"line"
string_set[  38]:   10029288415951054647 0x55b6f79c0ee8->"word"
string_set[  39]:   18413295556186398326 0x55b6f79b70b8->"putint"
string_set[  40]:   10319632053171904911 0x55b6f79b9978->"while"
string_set[  43]:    8627574202288022770 0x55b6f79bc2d8->"slot"
                    7372855250431704371 0x55b6f79ba598->"-"
string_set[  44]:      21606623943142546 0x55b6f79b8ab8->"{"
string_set[  45]:    6142509188972423790 0x55b6f79b6498->""
                    4687143246866533505 0x55b6f79b8518->"strcmp"
string_set[  47]:   13686721134187162920 0x55b6f79bb6a8->"}"
                   17605730733847049629 0x55b6f79b8ed8->"contin"
string_set[  48]:    6917069833844500161 0x55b6f79c05b8->"read_words"
string_set[  53]:   12051031859636400188 0x55b6f79b9e78->"]"
                   11659204968392427193 0x55b6f79b9448->"s2c"
string_set[  57]:    9386170080689500517 0x55b6f79b63d8->"expr"
string_set[  58]:    4017852018407011111 0x55b6f79b6658->"file"
                    3729804957429652673 0x55b6f79afd08->"<<ROOT>>"
string_set[  59]:   16579916199511856337 0x55b6f79b81f8->"getln"
string_set[  63]:    9611428142580219387 0x55b6f79b6158->"fail"
string_set[  64]:   12597418120160561823 0x55b6f79b67d8->"int"
string_set[  69]:    1873634190020599143 0x55b6f79bdb48->"else"
string_set[  70]:    9201800968376702500 0x55b6f79b7bf8->"getchr"
                   18169706682309360149 0x55b6f79b96b8->"cmp"
string_set[  71]:   14526079891305879776 0x55b6f79bbde8->">"
string_set[  76]:   15430489877420048373 0x55b6f79be238->"<="
string_set[  78]:    4706206