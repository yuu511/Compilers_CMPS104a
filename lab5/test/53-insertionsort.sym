getchr (6.28.1) {4} int function 

putstr (6.26.1) {3} void function 

putint (6.25.1) {2} void function 

fail (6.22.1) {0} void function 

main (7.72.1) {11} int function 
   count (7.73.4) {12} int variable local lval 0
   words (7.74.4) {12} string array variable local lval 1

getstr (6.29.1) {5} string function 

strcmp (7.8.1) {7} int function 
   s1 (7.8.13) {8} string variable param lval 0
   s2 (7.8.24) {8} string variable param lval 1
   index (7.9.4) {8} int variable local lval 2
   contin (7.10.4) {8} int variable local lval 3
   s1c (7.11.4) {8} int variable local lval 4
   s2c (7.12.4) {8} int variable local lval 5
   cmp (7.13.4) {8} int variable local lval 6

putchr (6.24.1) {1} void function 

getln (6.30.1) {6} string function 

insertion_sort (7.26.1) {8} void function 
   size (7.26.22) {9} int variable param lval 0
   words (7.26.32) {9} string array variable param lval 1
   sorted (7.27.4) {9} int variable local lval 2
   slot (7.28.4) {9} int variable local lval 3
   element (7.29.4) {9} string variable local lval 4
   contin (7.30.4) {9} int variable local lval 5

print_array (7.51.1) {9} void function 
   size (7.51.19) {10} int variable param lval 0
   words (7.51.29) {10} string array variable param lval 1
   index (7.52.4) {10} int variable local lval 2

read_words (7.60.1) {10} int function 
   size (7.60.17) {11} int variable param lval 0
   words (7.60.27) {11} string array variable param lval 1
   count (7.61.4) {11} int variable local lval 2
   word (7.62.4) {11} string variable local lval 3

