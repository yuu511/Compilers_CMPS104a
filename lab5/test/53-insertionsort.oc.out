CONDITIONAL WHILE STATEMENT SYMBOL TOK_IDENT 
CONDITIONAL IF STATEMENT SYMBOL TOK_NE 
CONDITIONAL IF STATEMENT SYMBOL TOK_EQ 
CONDITIONAL IF STATEMENT SYMBOL TOK_EQ 
CONDITIONAL WHILE STATEMENT SYMBOL TOK_LT 
CONDITIONAL WHILE STATEMENT SYMBOL TOK_IDENT 
CONDITIONAL IF STATEMENT SYMBOL TOK_EQ 
CONDITIONAL IF STATEMENT SYMBOL TOK_LE 
CONDITIONAL WHILE STATEMENT SYMBOL TOK_LT 
CONDITIONAL WHILE STATEMENT SYMBOL TOK_INTCON 
CONDITIONAL IF STATEMENT SYMBOL TOK_EQ 
CONDITIONAL IF STATEMENT SYMBOL TOK_EQ 
